### DISCUSSION TITLE
[PyTorch] How to clone an instance of Module?

### DISCUSSION THREAD

**Discussion by guangster** [bytedeco/javacpp-presets#1230] [Q&A] [answered]
How do you subclass Cloneable when extending Module defined in Java? (Or more generally how to properly clone a class defined in Java that extends Module?)

For example, in the SimpleMNIST example I am trying to clone `net`. However if I try to clone the net object directly here:
https://github.com/bytedeco/javacpp-presets/blob/master/pytorch/samples/SimpleMNIST.java#L38
by calling `net.clone()` I will get an error:
`clone() has not been implemented for torch::nn::Module. Subclass torch::nn::Cloneable<torch::nn::Module> instead of torch::nn::Module to inherit the ability to clone.`

The docs here http://bytedeco.org/javacpp-presets/pytorch/apidocs/org/bytedeco/pytorch/Module.html says
> ...Attempting to call the clone() method inherited from the base Module class (the one documented here) will fail. To inherit an actual implementation of clone(), you must subclass Cloneable. ...

which is what I'm guessing I have to extend instead of just the base Module. I'm guessing this is torch::nn::Cloneable, and not the java.lang's Cloneable interface (I tried that and it did not work). I saw Javacpp has code like https://github.com/bytedeco/javacpp-presets/blob/master/pytorch/src/gen/java/org/bytedeco/pytorch/Conv1dImplCloneable.java#L18 which then produces specific cloneable components. Can we do something like that in Java directly on an object defined using JavaCPP (like `net` above)?

**Comment by saudet** [comment]
That's a class template, so we can't create new instances of that without a C++ compiler. The easiest way to go about something like this is to fork the presets and add what you need as new files in `@Platform(include=...`, along with an updated `InfoMap`, in a similar fashion to what @yukoba did in the case of LLVM here:
https://github.com/bytedeco/javacpp-presets/blob/master/llvm/src/main/resources/org/bytedeco/llvm/include/FullOptimization.h

@HGuillemet might have other ideas.

**Comment by guangster** [comment]
I see. Are there any other ways to clone Module then? We would want to be able to define networks in Java without having to write C++ code every time.

**Reply by saudet** [reply]
@sbrunk Any ideas?

**Comment by HGuillemet** [comment, marked as answer]
If you'd like to clone an instance of a Java class subclassing Module, you need a Java implementation of `clone()`, since the C++ library knows nothing about the Java subclass. We could define some `CloneableModule` mapping `torch::nn::Cloneable<torch::nn::Module>` and Java modules could inherit from it instead of `Module`. So the Java implementation of `clone()` could call the C++ `clone()` from `Cloneable`, but it won't help since you'll get an object that is not an instance of the Java class.

For saving/loading models, I currently use `OutputArchive` and `Module.save`/`Module.load`.
Maybe could you use it for cloning too ?

**Reply by guangster** [reply]
Thanks for the ideas! I could certainly serialize then deserialize multiple times to achieve the desired effect.

Just out of curiosity, how would you implement `clone()` in Java (that is not calling the C++ `clone()` from `Cloneable`) though? Would you have to replicate the actual parameter Tensors manually (since we can't use `clone()` provided by e.g. LinearImpl)?

**Reply by guangster** [reply]
@saudet Following up on the serialize/deserialize idea, I see that JavaCPP's torch preset has this method in `OutputArchive` to write to a C stream. 
 http://bytedeco.org/javacpp-presets/pytorch/apidocs/org/bytedeco/pytorch/OutputArchive.html#save_to-org.bytedeco.javacpp.Pointer-

Is there a JavaCPP stream object I can pass here that can help me get some im-memory representation that I can then pass to
`InputArchive::load_from` so that I can do an in-memory clone?

Alternatively, I also see `save_to(...)` and `load_from(...)` accepting a `WriteFunction` and `ReadFunction` respectively, but I do not see these in the c++ torchlib's api. Is this relevant, and if so how can I use it?

**Reply by HGuillemet** [reply]
> Just out of curiosity, how would you implement `clone()` in Java (that is not calling the C++ `clone()` from `Cloneable`) though? Would you have to replicate the actual parameter Tensors manually (since we can't use `clone()` provided by e.g. LinearImpl)?

Instead of using the save/load feature, you could write a method that takes two instances of the same model, traverses recursively all the modules it consists in (calling `children()`), lists the registered parameters (`parameters()`), and possibly buffers (`buffers()`), of each module and `copy_()` their values.

**Reply by saudet** [reply]
ReadFunction and WriteFunction may not be documented, but it looks like they are easy to enough to implement in Java:
http://bytedeco.org/javacpp-presets/pytorch/apidocs/org/bytedeco/pytorch/presets/torch.ReadFunction.html
http://bytedeco.org/javacpp-presets/pytorch/apidocs/org/bytedeco/pytorch/presets/torch.WriteFunction.html
I'd try to implement that with a backing array or something and just write and read from that as per these 2 functions.

**Reply by guangster** [reply]
@saudet When you say implement in Java, do you mean implement a function that is a java version of the call method? The ReadFunction for example has a method `native long call(long pos, Pointer buf, long nbytes)`, so here we can instead have a Java function `public long call(...)` and work with the pos and buf objects in Java?

**Reply by saudet** [reply]
Yes, that's what it's for: https://github.com/bytedeco/javacpp#creating-callback-functions

**Reply by guangster** [reply]
In the example, it looks like the c code has to implement virtual function like `virtual void bar() {` and in Java I have to use the annotation `@Virtual`. However since ReadFunction is already defined with its own c header etc... I can't really replicate the example in your link.
Instead I tried 
```
        static class CloneWriter extends torch.WriteFunction {
            @Override
            @Virtual
            public native long call(Pointer buf, long nbytes);
        }
```

and then something like 
```
CloneWriter cloneWriter = new CloneWriter() {
                public long call(Pointer buf, long nbytes) {
                   ...
                }
            };
```
however even before calling `save_to(cloneWriter)` I get a crash. I tried some other options like just directly overriding call but they all seem to result in crash. Am I missing something? Or does this mean I would have to modify the c ReadFunction definition?

**Reply by saudet** [reply]
It's probably just getting garbage collected prematurely. Try to keep a reference in a field.

**Reply by saudet** [reply]
Also try with a direct subclass of WriterFunction, a subsubclass might not work indeed.

**Reply by HGuillemet** [reply]
Have you seen my comment above about copying values between 2 existing instances of your model, instead of using archives ? I believe it's a simpler solution.

**Reply by guangster** [reply]
@HGuillemet Thank you, I ended up going with this suggestion and calling `module.parameters(true)` and `module.buffers(true)` to get each `Tensor`s, then calling `tensor.copy_(...)` on them, it is working!

**Comment by saudet** [comment]
It's probably just getting garbage collected prematurely. Try to keep a
reference in a field.
